executor/aggregate.go

```diff
@@ -45,39 +96,484 @@ type HashAggExec struct {
+}
+
+func (e *HashAggExec) initForParallelExec() {
+	sessionVars := e.ctx.GetSessionVars()
+	finalConcurrency := sessionVars.HashAggFinalConcurrency
+	partialConcurrency := sessionVars.HashAggPartialConcurrency
+
+	e.isInputNull = true
+	e.finalOutputCh = make(chan *AfFinalResult, finalConcurrency)
+	e.inputCh = make(chan *HashAggInput, partialConcurrency)
+	e.finishCh = make(chan struct{}, 1)
+
+	e.partialOutputChs = make([]chan *HashAggIntermData, finalConcurrency)
+	for i := range e.partialOutputChs {
+		e.partialOutputChs[i] = make(chan *HashAggIntermData, partialConcurrency)
+	}
+	e.partialWorkers = make([]HashAggPartialWorker, partialConcurrency)
+	e.finalWorkers = make([]HashAggFinalWorker, finalConcurrency)
+	// Init partial workers.
+	for i := 0; i < partialConcurrency; i++ {
+		groupSets := make([]*mvmap.MVMap, finalConcurrency)
+		aggCtxsMappers := make([]aggCtxsMapper, finalConcurrency)
+		baseAggWorker := baseHashAggWorker{
+			finishCh:     e.finishCh,
+			aggFuncs:     e.AggFuncs,
+			maxChunkSize: e.maxChunkSize,
+			groupVals:    make([][]byte, 0, 8),
+		}
+
+		w := HashAggPartialWorker{
+			baseHashAggWorker: baseAggWorker,
+			inputCh:           make(chan *chunk.Chunk, 1),
+			outputChs:         e.partialOutputChs,
+			giveBackCh:        e.inputCh,
+			globalOutputCh:    e.finalOutputCh,
+			groupSets:         groupSets,
+			aggCtxsMaps:       aggCtxsMappers,
+			groupByItems:      e.GroupByItems,
+		}
+
+		e.partialWorkers[i] = w
+		e.inputCh <- &HashAggInput{
+			chk:        e.children[0].newChunk(),
+			giveBackCh: w.inputCh,
+		}
+	}
+
+	// Init final workers.
+	finalAggFuncs := e.newFinalAggFuncs()
+	for i := 0; i < finalConcurrency; i++ {
+		baseAggWorker := baseHashAggWorker{
+			finishCh:     e.finishCh,
+			aggFuncs:     finalAggFuncs,
+			maxChunkSize: e.maxChunkSize,
+			groupVals:    make([][]byte, 0, 8),
+		}
+		e.finalWorkers[i] = HashAggFinalWorker{
+			baseHashAggWorker:   baseAggWorker,
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupMap:            mvmap.NewMVMap(),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	var (
```

Comment:
Better not to define variables this way if we don't have to, it makes code ugly and somehow increases the code complexity.